### ISSUE TITLE
Component (openib) errors mlx4_0 device 

### ISSUE THREAD

**Issue by abeltre1** [open-mpi/ompi#5810]
## Background information
In this experiment, I am using Mellanox devices and drivers to run an MPI application (osu). I keep getting the errors below when I run with `openib` since the driver does not load for a specific host. When I take the particular host that `mpirun` throws as an error from the hostfile, it proceeds to throw another host, and another etc.  

### What version of Open MPI are you using? 

- mpirun (Open MPI) 3.1.1

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

- from Mellanox OFED software system distribution package 

### Please describe the system on which you are running

* Operating system/version: Centos 7.5
* Computer hardware: Intel Xion 

## Things are normal for TCP 
```shell
mpirun --mca mtl_mxm_np 0 --mca btl tcp,self --mca plm_rsh_no_tree_spawn 1 --map-by node -hostfile ~/hostfile -np 72 ./osu_alltoallv 
# OSU MPI All-to-Allv Personalized Exchange Latency Test v5.3.2
# Size       Avg Latency(us)
1                      43.25
2                      44.03
4                      44.47
8                      45.48
16                     47.70
32                     53.13
64                     64.61
128                   104.95
256                   178.48
512                   317.46
1024                  607.22
2048                  334.65
4096                  641.85
8192                 1624.47
16384                3257.20
32768                6246.35
65536               12574.22
131072              23627.27
262144              38249.50
524288              72622.47
1048576            139280.63
```
## Things are not normal for openib 
```bash 
mpirun --mca mtl_mxm_np 0 --mca btl self,openib --mca plm_rsh_no_tree_spawn 1 --map-by node -hostfile ~/hostfile -np 24 ./osu_alltoallv 


ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-4][[6526,1],3][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-5][[6526,1],22][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-5][[6526,1],4][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
--------------------------------------------------------------------------
WARNING: There was an error initializing an OpenFabrics device.

  Local host:   kube-infi-4
  Local device: mlx4_0
--------------------------------------------------------------------------
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-5][[6526,1],16][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-2][[6526,1],7][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-5][[6526,1],10][btl_openib_component.c:1670:init_one_device] ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-4][[6526,1],15][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-2][[6526,1],1][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-2][[6526,1],13][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-6][[6526,1],17][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
[kube-infi-2][[6526,1],19][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-1][[6526,1],6][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-1][[6526,1],12][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-1][[6526,1],0][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-6][[6526,1],5][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-4][[6526,1],9][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-1][[6526,1],18][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-6][[6526,1],11][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-4][[6526,1],21][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-6][[6526,1],23][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-3][[6526,1],8][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-3][[6526,1],14][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-3][[6526,1],20][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xffffffffffffffff valid_mask = 0x1)
[kube-infi-3][[6526,1],2][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument

# OSU MPI All-to-Allv Personalized Exchange Latency Test v5.3.2
# Size       Avg Latency(us)
1                      11.66
2                      11.40
4                      11.45
8                      11.43
16                     11.43
32                     11.61
64                     12.23
128                    13.16
256                    12.77
512                    14.50
1024                   21.37
2048                   38.99
4096                   69.41
8192                  158.51
16384                 296.76
32768                 566.62
65536                1176.81
131072               2343.93
262144               5253.02
524288              10401.06



```

**Comment by jladd-mlnx** [comment]
What version of MOFED are you running? Our recommendation is to use the UCX PML (`-mca pml ucx`). If you have a newer version of MOFED, then UCX will be the default PML in the MOFED packaged OMPI.

**Comment by abeltre1** [comment]
> What version of MOFED are you running? Our recommendation is to use the UCX PML (`-mca pml ucx`). If you have a newer version of MOFED, then UCX will be the default PML in the MOFED packaged OMPI.

MLNX_OFED
4.4-2.0.7.0

**Comment by mvpcaozixiang** [comment]
> > What version of MOFED are you running? Our recommendation is to use the UCX PML (`-mca pml ucx`). If you have a newer version of MOFED, then UCX will be the default PML in the MOFED packaged OMPI.
> 
> MLNX_OFED
> 4.4-2.0.7.0

I have the same problem,but I don't know how to view the version number of  MLNX_OFED.
Could you tell me ?
Have you solved this problem?
Thanks.

**Comment by abeltre1** [comment]
> > > What version of MOFED are you running? Our recommendation is to use the UCX PML (`-mca pml ucx`). If you have a newer version of MOFED, then UCX will be the default PML in the MOFED packaged OMPI.
> > 
> > 
> > MLNX_OFED
> > 4.4-2.0.7.0
> 
> I have the same problem,but I don't know how to view the version number of MLNX_OFED.
> Could you tell me ?
> Have you solved this problem?
> Thanks.

Hello @mvpcaozixiang , 

You can use `ofed_info` to collect all the packages installed in relation to MLNX_OFED.

**Comment by abeltre1** [comment]
> What version of MOFED are you running? Our recommendation is to use the UCX PML (`-mca pml ucx`). If you have a newer version of MOFED, then UCX will be the default PML in the MOFED packaged OMPI.

```
MLNX_OFED_LINUX-4.4-2.0.7.0 (OFED-4.4-2.0.7)
```

**Comment by abeltre1** [comment]
> What version of MOFED are you running? Our recommendation is to use the UCX PML (`-mca pml ucx`). If you have a newer version of MOFED, then UCX will be the default PML in the MOFED packaged OMPI.

@jsquyres  Do you know if anyone has looked into this particular issue?

**Comment by jsquyres** [comment]
@jladd-mlnx Can you please followup?

**Comment by kaonga** [comment]
As of 2018-10-11, we also see these same errors on:
**OS**: CentOS 7.5
**Mellanox OFED**: MLNX_OFED_LINUX-4.4-2.0.7.0, MLNX_OFED_LINUX-4.4-1.0.0.0, MLNX_OFED_LINUX-4.3-3.0.2.1
**OpenMPI**: 3.1.2
**Intel IMB Benchmarks**: 2018 Update 1

**Command/options**:
mpirun --mca btl openib,self,vader --mca pml ucx

**Comment by kaonga** [comment]
Hello all,

My message escapped before I completed my thoughts. I wanted to add the  log output to make it easier for those who may wish to see the error. The run starts with this complaint:

"ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0x27800000002 valid_mas
k = 0x1)"

and then runs until it gets to 8 processes where things just break down after 1024 byte packets have been sent and it hangs. The first error there is:

"mlx5: sm-node-02: got completion with error:
00000000 00000000 00000000 00000000
00000000 00000000 00000000 00000000
00000016 00000000 00000000 00000000
00000000 00008a12 0a000248 355849d2
[sm-node-01:5439 :0:5439] rc_verbs_iface.c:63   FATAL: send completion with erro
r: remote invalid request error
[sm-node-02:4036 :0:4036] rc_verbs_iface.c:63   FATAL: send completion with erro
r: remote invalid request error"

**Comment by abeltre1** [comment]
> processes where things just break down after 1024 byte packets have been sent and it hangs. The first error there is:
> 
> "mlx5: sm-node-02: got completion with error:
> 00000000 00000000 00000000 00000000
> 00000000 00000000 00000000 00000000

My OSU execution does not hang anywhere. It completes but has all the mentioned complaints/errors/device problems.

**Comment by abeltre1** [comment]
> Hello all,
> 
> Majd Dibbiny from Mellanox suggested that we do the following, which worked for us:
> 
> Add the following line:
> memset(&device->ib_exp_dev_attr, 0, sizeof(device->ib_exp_dev_attr));
> 
> at around line 1667 in the file
> /opt/ext_sw/openmpi-3.1.2/opal/mca/btl/openib/btl_openib_component.c
> 
> then rebuild Open MPI.

@kaonga that worked for me too! I tried it on 3.1.1

**Comment by jsquyres** [comment]
@hppritcha This seems like a simple solution.  Can you guys make a PR?

**Comment by angainor** [comment]
@jsquyres Just to jump in. I ran into the same issue. Any OpenMP version 3.0.0 and above fails to initialize openib with this message. The last one working is 2.1.5. 

@jladd-mlnx I am running MOFED 4.4.2, and I can use UCX pml. But if I understand correctly the problem is that openib is used for rdma osc module. So without it rdma will not work. And openshmem  doesn't work without openib either, because osc ucx doesn't work on ConnectX3.

**Comment by yosefe** [comment]
@angainor latest UCX master can work with osc/ucx on ConnectX-3

**Comment by hppritcha** [comment]
reopen to better track going in to other release branches

**Comment by justbennet** [comment]
Maybe the distribution tar ball at

https://download.open-mpi.org/release/open-mpi/v3.1/openmpi-3.1.3.tar.gz

did not get refreshed after this fix was implemented?  I downloaded that today, 22 Dec, and compiled and I get the warnings.

```
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xd8000000002 valid_mask = 0x1)
[bn01][[37143,17005],0][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0xd8100000002 valid_mask = 0x1)
[bn01][[37143,17005],1][btl_openib_component.c:1670:init_one_device] error obtaining device attributes for mlx4_0 errno says Invalid argument
--------------------------------------------------------------------------
WARNING: There was an error initializing an OpenFabrics device.

  Local host:   bn01
  Local device: mlx4_0
--------------------------------------------------------------------------
--------------------------------------------------------------------------
WARNING: There was an error initializing an OpenFabrics device.

  Local host:   bn01
  Local device: mlx4_0
--------------------------------------------------------------------------
```

It looks like Howard merged the fix on Dec 4, but the date listed for the 3.1.3 tarball on the open-mpi.org site is in Oct.

**Comment by jsquyres** [comment]
It was merged in to the v3.1.x branch after 3.1.3 was released -- it's slated for v3.1.4.

You can get a v3.1.4 pre-release snapshot here: https://www.open-mpi.org/nightly/v3.1.x/
